src/include/rho/GCNodeAllocator.hpp

```diff
@@ -0,0 +1,223 @@
+/*
+ *  R : A Computer Language for Statistical Data Analysis
+ *  Copyright (C) 1995, 1996  Robert Gentleman and Ross Ihaka
+ *  Copyright (C) 1998-2007   The R Development Core Team.
+ *  Copyright (C) 2008-2014  Andrew R. Runnalls.
+ *  Copyright (C) 2014 and onwards the Rho Project Authors.
+ *
+ *  Rho is not part of the R project, and bugs and other issues should
+ *  not be reported via r-bugs or other R project channels; instead refer
+ *  to the Rho website.
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, a copy is available at
+ *  https://www.R-project.org/Licenses/
+ */
+
+#ifndef RHO_GCNODEALLOCATOR_HPP
+#define RHO_GCNODEALLOCATOR_HPP
+
+#include <cstdint>
+#include <functional>
+
+#include "rho/AddressSanitizer.hpp"
+
+/** Reports an allocation error and calls abort(). */
+void allocerr(const char* message);
+
+namespace rho {
+  // Forward declarations.
+  class AllocationTable;
+  class AllocatorSuperblock;
+  class GCNode;
+
+  /**
+   * Freelists are used to store superblock nodes and large allocations.
+   * For large allocations the block and superblock members are unused.
+   */
+  struct FreeListNode {
```

Comment:
After looking into this more, I think using `boost::intrusive::slist` is difficult to integrate well with the quarantine list because it needs to unpoison the allocations in the freelist during iteration.